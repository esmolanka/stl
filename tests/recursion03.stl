
type Foo = < Nil, Foo : Foo >
type Bar = < Nil, Foo : Foo, Bar : Bar >

#check Bar <: Foo
#check Foo <: Bar

#check <Foo : Foo> <: Bar
#check <Foo : Foo> <: Foo

#check <Bar : Bar> <: Bar
#check <Bar : Bar> <: Foo

#check Bar <: <Foo : Bar>
#check Foo <: <Foo : Bar>

