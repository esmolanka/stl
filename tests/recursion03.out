tests/recursion03.stl:5:1-5:18:
  Subtype:
    μ Bar. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #). ∃ (ω : #).
      < Nil^ω : Unit
      | Foo^ω/1 :
          μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
      | Bar^ω/2 : Bar
      | ... ρ
      >
  Supertype:
    μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
  Result:
    FAIL
    Type is not polymorphic enough.
      Rigid variable ω does not unify with:
        ▫
tests/recursion03.stl:6:1-6:18:
  Subtype:
    μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
  Supertype:
    μ Bar. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #). ∃ (ω : #).
      < Nil^ω : Unit
      | Foo^ω/1 :
          μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
      | Bar^ω/2 : Bar
      | ... ρ
      >
  Result:
    OK
tests/recursion03.stl:8:1-8:26:
  Subtype:
    ∀ (ρ : Row). ∃ (ω : #).
      < Foo^ω :
          μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
      | ... ρ
      >
  Supertype:
    μ Bar. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #). ∃ (ω : #).
      < Nil^ω : Unit
      | Foo^ω/1 :
          μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
      | Bar^ω/2 : Bar
      | ... ρ
      >
  Result:
    OK
tests/recursion03.stl:9:1-9:26:
  Subtype:
    ∀ (ρ : Row). ∃ (ω : #).
      < Foo^ω :
          μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
      | ... ρ
      >
  Supertype:
    μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
  Result:
    OK
tests/recursion03.stl:11:1-11:26:
  Subtype:
    ∀ (ρ : Row). ∃ (ω : #).
      < Bar^ω :
          μ Bar. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #). ∃ (ω : #).
            < Nil^ω : Unit
            | Foo^ω/1 :
                μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
                  <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
            | Bar^ω/2 : Bar
            | ... ρ
            >
      | ... ρ
      >
  Supertype:
    μ Bar. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #). ∃ (ω : #).
      < Nil^ω : Unit
      | Foo^ω/1 :
          μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
      | Bar^ω/2 : Bar
      | ... ρ
      >
  Result:
    OK
tests/recursion03.stl:12:1-12:26:
  Subtype:
    ∀ (ρ : Row). ∃ (ω : #).
      < Bar^ω :
          μ Bar. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #). ∃ (ω : #).
            < Nil^ω : Unit
            | Foo^ω/1 :
                μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
                  <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
            | Bar^ω/2 : Bar
            | ... ρ
            >
      | ... ρ
      >
  Supertype:
    μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
  Result:
    FAIL
    Type is not polymorphic enough.
      Rigid variable ω does not unify with:
        ▫
tests/recursion03.stl:14:1-14:26:
  Subtype:
    μ Bar. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #). ∃ (ω : #).
      < Nil^ω : Unit
      | Foo^ω/1 :
          μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
      | Bar^ω/2 : Bar
      | ... ρ
      >
  Supertype:
    ∀ (ρ : Row). ∃ (ω : #).
      < Foo^ω :
          μ Bar. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #). ∃ (ω : #).
            < Nil^ω : Unit
            | Foo^ω/1 :
                μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
                  <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
            | Bar^ω/2 : Bar
            | ... ρ
            >
      | ... ρ
      >
  Result:
    FAIL
    Type is not polymorphic enough.
      Rigid variable ω does not unify with:
        ▫
tests/recursion03.stl:15:1-15:26:
  Subtype:
    μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
  Supertype:
    ∀ (ρ : Row). ∃ (ω : #).
      < Foo^ω :
          μ Bar. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #). ∃ (ω : #).
            < Nil^ω : Unit
            | Foo^ω/1 :
                μ Foo. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
                  <Nil^ω : Unit| Foo^ω/1 : Foo | ... ρ>
            | Bar^ω/2 : Bar
            | ... ρ
            >
      | ... ρ
      >
  Result:
    FAIL
    Type is not polymorphic enough.
      Rigid variable ω does not unify with:
        ▫